### DISCUSSION TITLE
Reasons to use MSEide +MSEgui instead of Lazarus + fpGUI?

### DISCUSSION THREAD

**Discussion by Wysardry** [mse-org/mseide-msegui#97] [Q&A]
I haven't created any programs with a graphical interface before, so I'm unsure what the differences are between using Lazarus + fpGUI compared with using MSEide + MSEgui.

Can anyone explain the advantages and disadvantages of both options?

If it helps, I would like to create programs along the lines of [Zettlr](https://www.zettlr.com/), [Obsidian](https://obsidian.md/) and [Scrivener](https://www.literatureandlatte.com/scrivener/).

Also, if I chose to use MSEide/MSEgui, would it be reasonably simple to do so with fpcupdeluxe (on Windows)? I'm not sure if the custom path settings for each install would be an issue.

**Comment by rchastain** [comment]
Using a compiler installed by fpcup is not an issue. You just have to configure MSEide.

MSEide is very powerful, but you will need time to discover how to use all its features. fpGUI has no native IDE. You can create fpGUI applications with a simple text editor like Notepad++. (This is how I made my [chess GUI](https://gitlab.com/rchastain/eschecs).)

Both solutions are good. I would say that fpGUI is maybe easier to learn; and MSEise/MSEgui maybe more sophisticated, and more powerful once you master it. With fpGUI, all is in code, whereas MSEide uses MFM files (created by the IDE).

By the way, MSEide needs contributors, people writing tutorials! Here is one [english tutorial](https://msegui.net/tutoriels/twidget-frame-face/tutoriel.html) made by me.

What kind of application do you intend to create?

**Reply by Wysardry** [reply]
Thanks. As I mentioned, I intend to create something along the lines of Zettlr, Obsidian and Scrivener.

In other words, a text editor that allows you to write and edit long documents split into separate smaller files that you can rearrange into any order and export as one file or many.

This is just an idea in my head so far, but I can see it (eventually) having multiple view windows, including a file tree view, a document tree view, a (markdown) text edit view, an HTML preview, file load/save/export dialogues, index card or other 2D document overview, metadata editor, colour coding for different text sections, colour themes (at least light and dark), spell checker and preferences editor(s).

I'm not sure at this point whether there will be tabbed windows, but there would defintely be multiple views on screen at the same time, which should be able to be independently resized, enabled and disabled. Being able to show the text editor in full screen (with menu) and distraction free mode (with no menues) would be a bonus.

I'm not keen on multiple floating windows, so the views would be "docked".

**Comment by graemeg** [comment]
For years I've used MSEide (the ide application) and developed fpGUI applications. Best of both worlds. :wink:  It's super fast (compared to Lazarus) and at the time, debugging was much nicer in MSEide. MSEide also made it really easy for me to switch between FPC versions and fpGUI versions - with a single click of the mouse.

![image](https://github.com/user-attachments/assets/37984e9d-2fb9-4371-ae19-a89962eeef06)

![image](https://github.com/user-attachments/assets/8f4dcd68-21d0-4e4a-b500-aeafbc7b0189)

![image](https://github.com/user-attachments/assets/10a837f5-7818-4198-947e-86cd1e95ec78)

**Reply by rchastain** [reply]
Great tutorial Graeme. Thank you!

**Reply by fredvs** [reply]
There is also the IDE ideU: https://github.com/fredvs/ideU/releases/tag/2.14.0
With fpGUI macros and fpGUI designer integrated.
(But there the integrated macro is ${FPGUIDIR} vs ${FPGUI})

**Reply by Wysardry** [reply]
Thanks. That looks like it would make the process much easier.

Is there any reason why switching between versions wouldn't also work with those installed by fpcupdeluxe? As I understand it, the only difference is that you have to use the special file paths it sets up for you.

**Reply by Wysardry** [reply]
> There is also the IDE ideU: https://github.com/fredvs/ideU/releases/tag/2.14.0 With fpGUI macros and fpGUI designer integrated. (But there the integrated macro is ${FPGUIDIR} vs ${FPGUI})

Do you have a page with more information about ideU? There doesn't seem to be any info in the wiki or a link to a website on the project page.

**Reply by fredvs** [reply]
ideU is a fork of MSEide.
There are 2 distinct things in our discussion:
- The IDE (Integrated Development Environment) who gives a source editor, debugging feature, some a graphic form editor (including Palette of components, Object Inspector, ...), compilation feature.
 Lazarus, MSEide and ideU are IDE.

- The Graphic Toolkit (widget-set) with many graphic components, ready to use, like forms, buttons, grids, etc.  LCL, fpGUI and MSEgui are widget-sets.

Lazarus (the IDE) has his own widget-set, called LCL that uses third party C widget-sets, like GTK, Qt, Carbon, ... with the LCL form designer + palette of components + object inspector.

MSEide (shared with ideU) has his own widget-set too, called MSEgui, that does not use any dependencies and also has his own form designer + palette of components + object inspector.
Source using MSEgui widgetset can also be used with Lazarus IDE but the LCL form designer is not compatible with MSEgui designer, so only debugging, editing source and compiling can be done.

fpGUI is a widget-set without any dependencies that can be used with Lazarus or MSEide-ideU and has his own independent form designer: UIDesigner.

For the infos about ideU, there are some how to install on the release page but mainly all what is explained for MSEide is valable for ideU.

**Reply by graemeg** [reply]
I used that setup long before fpcupdeluxe existed, so that was just what I was used to. Also the screenshot are from a very simple project. Mist of my other projects where a lot more complex in project dependencies, and I could control any combination directly from inside the ide.

Obviously my projects never depended on any single ide. I could switch to Lazarus IDE too, or build with FPC built-in package tool from the command line. MSEide was simply my most frequent used ide, for the reasons I mentioned before.
